output/formatter.go

```diff
@@ -180,27 +180,36 @@ func convertToSonarIssues(rootPaths []string, data *reportInfo) (*sonarIssues, e
 func convertToSarifReport(rootPaths []string, data *reportInfo) (*sarifReport, error) {
 	sr := buildSarifReport()
 
-	var rules []*sarifRule
-	var locations []*sarifLocation
+	rules := make([]*sarifRule, 0)
 	results := []*sarifResult{}
 
-	for index, issue := range data.Issues {
-		rules = append(rules, buildSarifRule(issue))
+	for _, issue := range data.Issues {
```

Comment:
This approach doesn't seem right to me because it is removing issues form the result produced by the same rule but at a different location. 

It seems that the recommended way to solve this in the spec (http://docs.oasis-open.org/sarif/sarif/v2.0/csprd01/sarif-v2.0-csprd01.html - see section `3.19.6 ruleId property`) is to append an index to to the duplicated rule id. In our case something like: 

```
G101 (CEW-123)
G101 (CEW-123)-1
G101 (CEW-123)-2
G101 (CEW-123)-3
....
```

Maybe the best way to solve this is to extend the signature of `buildSafariRule` function with an index like this:

```
func buildSarifRule(issue *gosec.Issue, index int) *sarifRule 
```

We can skip appending the index `0` to rule ID because is the first encountered rule but start appending it form `1`. 

Also I think that you can keep track of duplicated rules in `convertToSafariReport` function by using a map instead of iterating twice over the rules slice:

```
rules := []*safariRule{}
rulesMap := map[string]int


for index, issue := range data.Issues {

      ruleIdIndex, ok := rulesMap[issue.RuleID]; 
      if ok {
           ruleidIndex += 1
            rulesMap[issue.RuleID] = ruleIndex
      } else {
           rulesMap[issue.RuleID] = 0
      }

       rules = append(rules, buildSafariRule(issue, ruleidIndex))
....
```

Reply:
>  it is removing issues from the result produced by the same rule but at a different location.
I don't think it's true, but maybe I misunderstand you. Let me explain what I see happens in the report:
- we add deduplicated rules
- we add all rule violations one-by-one, each violation refers to the rule and the location

Here's the excerpt from the report:
```
{
	"runs": [
		{
			"tool": {
				"driver": {
					"rules": [
						{
							"id": "G404 (CWE-338)",
							...
						},
						{
							"id": "G402 (CWE-295)",
							...
						}
					]
				}
			},
			"results": [
				{
					"ruleId": "G404 (CWE-338)",
					"ruleIndex": 0,
					"locations": [
						{
							"physicalLocation": {
								"artifactLocation": {
									"uri": "api/chisel/tunnel.go"
								},
								...
							}
						}
					]
				},
				{
					"ruleId": "G404 (CWE-338)",
					"ruleIndex": 0,
					"locations": [
						{
							"physicalLocation": {
								"artifactLocation": {
									"uri": "api/chisel/proxy.go"
								},
								...
							}
						}
					]
				},
				{
					"ruleId": "G402 (CWE-295)",
					"ruleIndex": 1,
					"locations": [
						{
							"physicalLocation": {
								"artifactLocation": {
									"uri": "api/http/handler/websocket/proxy.go"
								},
								...
							}
						}
					]
				},
				...
			]
		}
	]
}

```
You can see 2 result items that have the same ruleId `G404 (CWE-338)` and different locations.

As for the dedup mechanics, I surely can re-write it using a map.

Reply:
Thanks for this clarification and for improving the logic by using a map.